### ISSUE TITLE
`zod-form-data` Checkbox union reversed?

### ISSUE THREAD

**Issue by TheRealFlyingCoder** [airjp73/remix-validated-form#37]
Hopefully i'm just running on empty atm (it's 5pm), but it seems the default `zfd.checkbox()` allows undefined, but not checked?

With the following, I can submit unchecked, but fails validation on checked:

```
import { withZod } from '@remix-validated-form/with-zod';
import { useField, ValidatedForm } from 'remix-validated-form';
import { z } from 'zod';
import { zfd } from 'zod-form-data';

const validator = withZod(
	z.object({
		myCheckbox: zfd.checkbox(),
	}),
);

const Checkbox: React.FC<{ name: string, label: string}> = ({ name, label }) => {
    const { error, getInputProps } = useField(name);

    return (
        <div>
            <label className="flex items-center">
                <input
	            {...getInputProps({
                        id: name,
                        type: 'checkbox',
                    })}
                />
               {label}
            </label>
            {error && (
                <span>
	            {error}
                </span>
            )}
        </div>
    )
};

export default () => (
	<ValidatedForm validator={validator} method="post">
		<Checkbox label="Checkbox" name="myCheckbox" />
		<button type="submit">Submit</button>
	</ValidatedForm>
);
```

By default this checkbox should be required, with `zfd.checkbox().optional()` to allow undefined right?

**Comment by airjp73** [comment]
Had a little time to take look and reproduced the issue. My reproduction is a little different from your example though so a reproduction from you could still be helpful. This appears to happen when `value={undefined}` is set on the checkbox input. This results in the checked state being an empty string instead of `"on"`.

This issue also came up in #34, but got lost in the amount of stuff going on there.

I had this issue in my inputs in the test app, too. I was doing
```tsx
<input
  {...getInputProps({
    type,
    ref,
    id: name,
    value
  })}
/>
```
and `value` was an optional prop.

The fix is to completely omit the `value` prop rather than set it to `undefined`. This is super straightforward to handle in `getInputProps`, but I can't help but wonder if the "correct" fix is to instead educate about how this works in the docs and have the prop be omitted in userland. I'm open to thoughts/opinions on this.

**Comment by TheRealFlyingCoder** [comment]
>The fix is to completely omit the value prop rather than set it to undefined.

I'm not sure I follow, in my code `value` is omitted from `getInputProps()` but the issue still occurs.

In my mind, empty string is still counted checked, as the form submission will omit any unchecked fields by default, resulting in undefined.

The fix for that might simply be going:

```
export const checkbox = ({ trueValue }: CheckboxOpts = {}) =>
  z.union([
    z.string().refine(value => !!trueValue ? value === trueValue : value !== undefined).transform(() => true),
    z.literal(undefined).transform(() => false),
  ]);
```  

If true value is passed in, the user has done `getInputProps({ value: 'myValue' })` so you match it, otherwise just ensure the value is a string (regardless of whats in it)?

**Comment by airjp73** [comment]
> I'm not sure I follow, in my code value is omitted from getInputProps() but the issue still occurs.

If you could provide a reproduction of this, that would be really helpful 😃. I would like to figure out why that's happening and take that into account when making a decision about how to fix.

> In my mind, empty string is still counted checked, as the form submission will omit any unchecked fields by default, resulting in undefined.

I went back and forth on this when writing the library. I think the approach you describe is valid, but I wanted to err on the side of strictness. If I built the form, then I know exactly what value to expect, and for the majority of use-cases that's going to be `on` because you only need a `value` for checkbox groups.

**Comment by airjp73** [comment]
@TheRealFlyingCoder, I think I realized what the issue is. You posted your issue before I released 3.4, and that's the release where I fixed checkbox inputs. The issue you're having is probably the same as the one in #34. ([release for 3.4](https://github.com/airjp73/remix-validated-form/releases/tag/remix-validated-form-v3.4.0)).

With the latest version, I copy-pasted your code into a route and it's working.
